js/src/tags.js

```diff
@@ -0,0 +1,467 @@
+        return item;
+    };
+
+    ListableItemsAdmin.prototype.getModifiedItem = function (id) {
+        if (!(id in this.modified_items)) {
+            this.modified_items[id] = JSON.parse(JSON.stringify(this.loaded_items[id]));
+        }
+        return this.modified_items[id];
+    };
+
+    ListableItemsAdmin.prototype.updateItemFromForm = function (id) {
+        var row = $('.' + this.row_class + '[data-id=' + id + ']');
+        var item = this.getModifiedItem(id);
+
+        this.fillItemFromRow(row, item);
+
+        if (_.isEqual(item, this.loaded_items[id])) {
+            this.resetEdit(id);
+        }
+        this.refreshDisplay();
+    };
+
+    ListableItemsAdmin.prototype.resetAll = function () {
+
+        var offset;
+        for (offset in this.new_item_ids) {
+            this.forgetItem(this.new_item_ids[offset]);
+        }
+
+        for (var id in this.modified_items) {
+            this.resetEdit(id);
+        }
+
+        for (offset in this.deleted_items_ids) {
+            this.resetEdit(this.deleted_items_ids[offset]);
+        }
+
+        this.refreshDisplay();
+    };
+
+    ListableItemsAdmin.prototype.resetEdit = function (id) {
+        delete this.modified_items[id];
+        delete this.item_errors[id];
+
+        if (this.deleted_items_ids.includes(id)) {
+            this.deleted_items_ids.splice(this.deleted_items_ids.indexOf(id));
+        }
+        if (this.new_item_ids.includes(id)) {
+            this.getModifiedItem(id);
+        }
+    };
+
+    ListableItemsAdmin.prototype.save = function () {
+        var self = this;
+
+        self.item_errors = {};
+
+        var new_item_ids = self.new_item_ids;
+        var modified_items = self.modified_items;
+        var url = self.saveUrl;
+
+        var body = {
+            created: {},
+            modified: {},
+            deleted: self.deleted_items_ids
+        };
+
+        for (var id in modified_items) {
+            if (new_item_ids.includes(id)) {
+                body.created[id] = modified_items[id];
+            } else {
+                body.modified[id] = modified_items[id];
+            }
+        }
+
+        $.ajax({
+            url: url,
+            type: 'POST',
+            dataType: 'json',
+            contentType: 'application/json',
+            data: JSON.stringify(body)
+        }).done(function (response) {
+            if ('errors' in response) {
+                self.item_errors = response.errors;
+            }
+
+            for (var id in self.loaded_items) {
+                if (!(id in self.item_errors)) {
+                    self.forgetItem(id);
+                }
+            }
+
+            self.changePage(1);
+            self.loadPageCount();
+        });
+    };
+
+    ListableItemsAdmin.prototype.refreshDisplay = function () {
+        this.items_list.empty();
+
+        for (var offset in this.column_headings) {
+            var heading = $($.parseHTML(_.template(this.template_heading.html())({})));
```

Comment:
Querying the DOM and compiling the template (`_.template()` call) shouldn't be necessary this many times. As with my comment on `modal.html`, you should be able to compile the template once and use the result as a function.

Also, like the other comment, you can use placeholders in these [underscore templates](https://underscorejs.org/#template).

**Note:** our template placeholders use a different syntax than the default. See `src/js/underscore.js` in Core for the regexes. Interpolation is just value substitution, evaluation is actual code evaluation, and escaping is for any user input text.

Reply:
Fixed in ef1176accf5c411a39af7103186807ffe773720a